---
title: "phenopath_campbell.Rmd"
author: "Kevin Stachelek"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r , include=FALSE}
library(stchlk.notes)
pwd <- getwd()
stchlk.notes::loadkey(pwd)

library(tidyverse)
library(phenopath)

```

* synamic behavior should be studied in logitudinal framework. 

* pt methods rely on the assumption that any two individuals should carry correspondingly similar pts. 

* pt typically captures some variation that corresponds to continuous change of a set of biological pathways (no branching)

* pt can be used with single cell data to capture:
  + cell cycle
  + branching 
  
* pt suffers from heritage in modeling cancer (progressive diseases)

* with phenopath can answer:
  +interaction between hetero in external factors and biological progression

# Methods


# Simulate Data

```{r}
set.seed(123L)
sim <- simulate_phenopath()
```

```{r}
print(str(sim))
```

```{r}
df_gex <- tbl_df(sim$y[,c(1,3,11,13,21,23,31,33)]) %>% 
  mutate(x = factor(sim[['x']]), z = sim[['z']]) %>% 
  gather(gene, expression, -x, -z) %>% 
  identity()

ggplot(df_gex, aes(x=z, y = expression, color = x)) +
  geom_point() +
  facet_wrap(~ gene, nrow = 2) +
  scale_color_brewer(palette = "Set1")
```

plot in pca space. color by covariate and pt

```{r}
pca_df <- tbl_df(prcomp(sim$y)$x[,1:2]) %>% 
  mutate(x = factor(sim[['x']]), z = sim[['z']])

ggplot(pca_df, aes(x = PC1, y = PC2, color = x)) +
  geom_point() + 
  scale_colour_brewer(palette = "Set1")

ggplot(pca_df, aes(x = PC1, y = PC2, color = z)) +
  geom_point()
```

fit phenopath model

requires an expression matrix `y` and covariate vector `x`

expression data should be logged counts such as log2(tpm+1)

```{r}
fit <- phenopath(sim$y, sim$x, elbo_tol = 1e-6, thin = 40)

print(fit)

plot_elbo(fit)
```

examine results

```{r}
qplot(sim$z, trajectory(fit)) +
  xlab("True z") + 
  ylab("Phenopath z") 

qplot(sim$z, pca_df$PC1) + 
  xlab("True z") + 
  ylab("PC1")
```

check correlation of phenopath predicted pt wiht true pt

```{r}
cor(sim$z, trajectory(fit))
```

find interactions between latent space and covariates use three functions:
* `interaction effects`
* `interaction_sds`
* `significant_interactions`

or can use the `interactions` function to get results as df 


```{r}
gene_names <- paste0("gene", seq_len(ncol(fit$m_beta)))
df_beta <- data_frame(beta = interaction_effects(fit), beta_sd = interaction_sds(fit), is_sig = significant_interactions(fit), gene = gene_names)

df_beta$gene <- fct_relevel(df_beta$gene, gene_names)

ggplot(df_beta, aes(x = gene, y = beta, color = is_sig)) +
  geom_point() + 
  geom_errorbar(aes(ymin = beta - 2 * beta_sd, ymax = beta + 2* beta_sd)) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1),
        axis.title.x = element_blank()) +
  ylab(expression(beta)) +
  scale_color_brewer(palette = "Set2", name = "Significant")
```

first 20 plotted without interaction effects, next 20 with. graph largest effect size now:

```{r}
which_largest <- which.max(df_beta$beta)

df_large <- data_frame(y = sim[['y']][, which_largest],
                       x = factor(sim[['x']]),
                       z = sim[['z']]
                       )

ggplot(df_large, aes(x = z, y = y, color = x)) + 
  geom_point() +
  scale_color_brewer(palette = "Set1") + 
  stat_smooth()
```




# using summarizedexperiment input 

PhenoPath will use by default is in the exprs assay of a SummarizedExperiment (ie assay(sce, "exprs")) as gene expression values, which can be changed using the sce_assay string in the column to phenopath.

```{r}
suppressPackageStartupMessages(library(SummarizedExperiment))
exprs_mat <- t(sim$y)
pdata <- data.frame(x = sim$x)
sce <- SummarizedExperiment(assays = list(exprs = exprs_mat), 
                            colData = pdata)
sce
```

can pass covariates to phenopath either:
* as vector or matrix
* as character that names column in `colData(sce)`
* formula to build a model matrix from `colData(sce)`

```{r}
fit <- phenopath(sce, sim$x) # 1
# fit <- phenopath(sce, "x") # 2
# fit <- phenopath(sce, ~ x) # 3
```


# initialize latent space 

can be done either through: 

* An integer specifying a principal component of the data to initialise to
* A vector specifying the initial values
* Random initialisation from standard normal distribution


```{r}
fit <- phenopath(sim$y, sim$x, z_init = 1) # 1, initialise to first principal component
# fit <- phenopath(sim$y, sim$x, z_init = sim$z) # 2, initialise to true values
# fit <- phenopath(sim$y, sim$x, z_init = "random") # 3, random initialisation
```

# control variational inference 

```{r}
fit <- phenopath(sim$y, sim$x,
                 maxiter = 1000, # 1000 iterations max
                 elbo_tol = 1e-2, # consider model converged when change in ELBO < 0.02%
                 thin = 20 # calculate ELBO every 20 iterations
                 )
```





* Bayesian framework for pt traj modeling allows explicit inclusion of prior factors. 
* incorporate informatio nas covariates that can change pt, allows subgroup to develop own trajs. 

* combines linear regression and latent variable modeling allows for interactions between covariates and time components of model. 

# Results

* objective is prob ordering of high-D expression data.

* obj achieved by compressing information in data to one dimension (pt)

* observed data for nth individual is a linear function of measuered covariates and unobserved latent 

