---
title: "tscan_tutorial.Rmd"
author: "Kevin Stachelek"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r , include=FALSE}

library(TSCAN)
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86

lookup_genes <- function(txname){
  
  txs <- transcripts(edb, filter = TxIdFilter(txname),
                     columns = c("symbol"))
  return(txs$symbol)
  
}

lookup_transcripts <- function(tx_df){
  
  txids <- rownames(tx_df)
  
  txs <- transcripts(edb, filter = TxIdFilter(txids), columns = c("symbol"))
  
  names(txids) <- txs$symbol
  
  return(txids)
}

```



```{r}
#data(lpsdata)

#combining the raw count matrix of all 3 fetal datasets
#df1<-read_0407 %>% 
 #inner_join(read_1001,by=c("XX1")) %>% 
 #inner_join(read_1031,by=c("XX1"))

write_csv(df1,"~/single_cell_projects/quicklinks/3_seq_dshayler_proj/output/all_fetal_transcript_count_matrix.csv")

shlpath <- "~/single_cell_projects/quicklinks/3_seq_dshayler_proj/output/all_fetal_transcript_count_matrix.csv"
shldata <- read.csv(shlpath, row.names = 1)

###### filter out genes that are not expressed in most cells
procshl <- preprocess(shldata)


# procdata <- preprocess(lpsdata)
procdata <- procshl


```

```{r}

#### Dimensional reduction using PCA and model based clustering

cluster_info<-read.csv("~/single_cell_projects/quicklinks/3_seq_dshayler_proj/cluster_info.csv")

cluster_vector<-as.numeric(cluster_info[[2]]) #convert the df into vector
names(cluster_vector)<-cluster_info$SampleID


#trace("exprmclust", quote(browser(skipCalls = 4)),
     # exit = quote(browser(skipCalls = 4)))



lpsmclust <- exprmclust(procdata, clusternum=1:4, cluster = cluster_vector)

plotmclust(lpsmclust)
```


```{r}
lpsorder <- TSCANorder(lpsmclust)
lpsorder

write.csv(lpsorder,"~/single_cell_projects/quicklinks/3_seq_dshayler_proj/results/pseudotime_table_all_fetal_cells.csv")
```

```{r}
diffval <- difftest(procdata,lpsorder)
#Selected differentially expressed genes under qvlue cutoff of 0.05
head(row.names(diffval)[diffval$qval < 0.05])
```

```{r}
# log2 transform data for a given gene

get_symbol_expr_vec <- function(symbol, df){
  # browser()
  txs <- lookup_transcripts(df)
  txs <- txs[names(txs) == symbol]
  trx_sum <- colSums(df[txs,])
  return(trx_sum)
}

STAT2expr <- log2(get_symbol_expr_vec("RXRG", shldata)+1)


# STAT2expr <- log2(lpsdata["STAT2",]+1)

# trace("singlegeneplot", quote(browser(skipCalls = 4)),
#       exit = quote(browser(skipCalls = 4)))

singlegeneplot(STAT2expr, TSCANorder(lpsmclust,flip=TRUE,orderonly=FALSE))
```







