<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Single cell analysis package: sc-analyses Namespace Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Single cell analysis package
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#func-members">Functions</a> &#124;
<a href="#var-members">Variables</a>  </div>
  <div class="headertitle">
<div class="title">sc-analyses Namespace Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classsc-analyses_1_1settings.html">settings</a></td></tr>
<tr class="memdesc:"><td class="mdescLeft">&#160;</td><td class="mdescRight">this class reads settings of the run and keeps them in its attributes if settings file in incorrect, the init method prints error and terminates application  <a href="classsc-analyses_1_1settings.html#details">More...</a><br /></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="func-members"></a>
Functions</h2></td></tr>
<tr class="memitem:a96e3d0a6c4cbd66f450cd13d0597d693"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a96e3d0a6c4cbd66f450cd13d0597d693">read_expression</a> (expression_file, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>, min_expression=0.1, min_cells=10, log_transform=True)</td></tr>
<tr class="memdesc:a96e3d0a6c4cbd66f450cd13d0597d693"><td class="mdescLeft">&#160;</td><td class="mdescRight">function takes expression file and settings object and returns:  <a href="#a96e3d0a6c4cbd66f450cd13d0597d693">More...</a><br /></td></tr>
<tr class="separator:a96e3d0a6c4cbd66f450cd13d0597d693"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8c5c34b3ded1400d6294ac2263f02c52"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a8c5c34b3ded1400d6294ac2263f02c52">run_PCA</a> (expression_table, annotation, n_components)</td></tr>
<tr class="memdesc:a8c5c34b3ded1400d6294ac2263f02c52"><td class="mdescLeft">&#160;</td><td class="mdescRight">runs PCA and returns:  <a href="#a8c5c34b3ded1400d6294ac2263f02c52">More...</a><br /></td></tr>
<tr class="separator:a8c5c34b3ded1400d6294ac2263f02c52"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa11b06d3030a58a0287811a14e00de0c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#aa11b06d3030a58a0287811a14e00de0c">get_isoforms_correlated_with_pc</a> (expression_table, pc, n, filename)</td></tr>
<tr class="memdesc:aa11b06d3030a58a0287811a14e00de0c"><td class="mdescLeft">&#160;</td><td class="mdescRight">save genes correlated with a PC to file  <a href="#aa11b06d3030a58a0287811a14e00de0c">More...</a><br /></td></tr>
<tr class="separator:aa11b06d3030a58a0287811a14e00de0c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abca3b5ef9a0cde5e39c94287e35b31da"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#abca3b5ef9a0cde5e39c94287e35b31da">annotate_df</a> (row, df, min_dist, ax)</td></tr>
<tr class="memdesc:abca3b5ef9a0cde5e39c94287e35b31da"><td class="mdescLeft">&#160;</td><td class="mdescRight">create annotation label for a point on axis if it's far enough from other points used internally by plotting functions  <a href="#abca3b5ef9a0cde5e39c94287e35b31da">More...</a><br /></td></tr>
<tr class="separator:abca3b5ef9a0cde5e39c94287e35b31da"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af8827644de4784f9b91bd82a316ba4c7"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#af8827644de4784f9b91bd82a316ba4c7">plot_2d_pca_multiplot</a> (transformed_expression, annotation, pca, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>)</td></tr>
<tr class="memdesc:af8827644de4784f9b91bd82a316ba4c7"><td class="mdescLeft">&#160;</td><td class="mdescRight">create plot of 6 PC combinations PC1 vs PC2, PC3 vs PC4 etc.  <a href="#af8827644de4784f9b91bd82a316ba4c7">More...</a><br /></td></tr>
<tr class="separator:af8827644de4784f9b91bd82a316ba4c7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abbc100d6dc158e3b5a1fbba583da3e10"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#abbc100d6dc158e3b5a1fbba583da3e10">plot_2d_pca_single_plot</a> (transformed_expression, annotation, pca, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>)</td></tr>
<tr class="memdesc:abbc100d6dc158e3b5a1fbba583da3e10"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot cells of defined pair of PCs arguments are:  <a href="#abbc100d6dc158e3b5a1fbba583da3e10">More...</a><br /></td></tr>
<tr class="separator:abbc100d6dc158e3b5a1fbba583da3e10"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7ce99634f6c5778fe1ffb3a6333840f8"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a7ce99634f6c5778fe1ffb3a6333840f8">plot_3d_pca</a> (transformed_expression, annotation, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>, height=1080, width=1600)</td></tr>
<tr class="memdesc:a7ce99634f6c5778fe1ffb3a6333840f8"><td class="mdescLeft">&#160;</td><td class="mdescRight">create 3d PCA plot using plotly library arguments are:  <a href="#a7ce99634f6c5778fe1ffb3a6333840f8">More...</a><br /></td></tr>
<tr class="separator:a7ce99634f6c5778fe1ffb3a6333840f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae3c71f2d759657520fbffb935d9f727c"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#ae3c71f2d759657520fbffb935d9f727c">plot_hierarchycal_clusterings</a> (transformed_expression, annotation, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>)</td></tr>
<tr class="memdesc:ae3c71f2d759657520fbffb935d9f727c"><td class="mdescLeft">&#160;</td><td class="mdescRight">plot hierarchycal clustering arguments are:  <a href="#ae3c71f2d759657520fbffb935d9f727c">More...</a><br /></td></tr>
<tr class="separator:ae3c71f2d759657520fbffb935d9f727c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a48155c82bbcbbab0bf4a6b4ae399252d"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a48155c82bbcbbab0bf4a6b4ae399252d">rotate_expression</a> (transformed_expression, x, y, angle)</td></tr>
<tr class="memdesc:a48155c82bbcbbab0bf4a6b4ae399252d"><td class="mdescLeft">&#160;</td><td class="mdescRight">rotate transformed expression matrix by defined angle used internally in order to define pseudotime arguments are:  <a href="#a48155c82bbcbbab0bf4a6b4ae399252d">More...</a><br /></td></tr>
<tr class="separator:a48155c82bbcbbab0bf4a6b4ae399252d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f980392d707316c84323ff4da652e62"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a8f980392d707316c84323ff4da652e62">find_pseudotime</a> (transformed_expression, annotation, pca, <a class="el" href="classsc-analyses_1_1settings.html">settings</a>)</td></tr>
<tr class="memdesc:a8f980392d707316c84323ff4da652e62"><td class="mdescLeft">&#160;</td><td class="mdescRight">function  <a href="#a8f980392d707316c84323ff4da652e62">More...</a><br /></td></tr>
<tr class="separator:a8f980392d707316c84323ff4da652e62"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f5f78fd44190040bfde554a412bceaa"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a8f5f78fd44190040bfde554a412bceaa">plot_gene_with_pseudotime</a> (exp, pseudotime, transcript_id)</td></tr>
<tr class="memdesc:a8f5f78fd44190040bfde554a412bceaa"><td class="mdescLeft">&#160;</td><td class="mdescRight">plots gene expression over pseudotime arguments are:  <a href="#a8f5f78fd44190040bfde554a412bceaa">More...</a><br /></td></tr>
<tr class="separator:a8f5f78fd44190040bfde554a412bceaa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2dd8534ee3983dc13dac25891db005fe"><td class="memItemLeft" align="right" valign="top">def&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a2dd8534ee3983dc13dac25891db005fe">main</a> ()</td></tr>
<tr class="memdesc:a2dd8534ee3983dc13dac25891db005fe"><td class="mdescLeft">&#160;</td><td class="mdescRight">main function when run separately, program expects following arguments:  <a href="#a2dd8534ee3983dc13dac25891db005fe">More...</a><br /></td></tr>
<tr class="separator:a2dd8534ee3983dc13dac25891db005fe"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="var-members"></a>
Variables</h2></td></tr>
<tr class="memitem:a5c4820a68409c2ca9d5965d48c1643c8"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a5c4820a68409c2ca9d5965d48c1643c8">run_modes</a> = [&quot;find_day_correlated_pcs&quot;, &quot;2d-pca-multiplot&quot;, &quot;2d-pca-single&quot;, &quot;3d-pca&quot;, &quot;hierarchy&quot;, &quot;pseudotime&quot;]</td></tr>
<tr class="memdesc:a5c4820a68409c2ca9d5965d48c1643c8"><td class="mdescLeft">&#160;</td><td class="mdescRight">what modes can be script run in  <a href="#a5c4820a68409c2ca9d5965d48c1643c8">More...</a><br /></td></tr>
<tr class="separator:a5c4820a68409c2ca9d5965d48c1643c8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a43a788e866dfeecd5963f4ee948d4008"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#a43a788e866dfeecd5963f4ee948d4008">accepted_sets_with_parameter</a> = [&quot;color&quot;, &quot;outline-color&quot;, &quot;size&quot;, &quot;name&quot;, &quot;shape&quot;, &quot;superimpose&quot;, &quot;superimpose-for-spearman&quot;]</td></tr>
<tr class="memdesc:a43a788e866dfeecd5963f4ee948d4008"><td class="mdescLeft">&#160;</td><td class="mdescRight">sets with parameter look like: operation set_name parameter for ex.  <a href="#a43a788e866dfeecd5963f4ee948d4008">More...</a><br /></td></tr>
<tr class="separator:a43a788e866dfeecd5963f4ee948d4008"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af635bfeee6b34ad984fe2512c83d6be3"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#af635bfeee6b34ad984fe2512c83d6be3">accepted_sets_without_parameter</a> = [&quot;remove&quot;]</td></tr>
<tr class="memdesc:af635bfeee6b34ad984fe2512c83d6be3"><td class="mdescLeft">&#160;</td><td class="mdescRight">sets without parameter look like: operation set_name for ex.  <a href="#af635bfeee6b34ad984fe2512c83d6be3">More...</a><br /></td></tr>
<tr class="separator:af635bfeee6b34ad984fe2512c83d6be3"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adb95e5cb1cbb2c43a1889131cb017c2b"><td class="memItemLeft" align="right" valign="top">list&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="namespacesc-analyses.html#adb95e5cb1cbb2c43a1889131cb017c2b">accepted_parameters</a> = [&quot;number_of_genes&quot;]</td></tr>
<tr class="memdesc:adb95e5cb1cbb2c43a1889131cb017c2b"><td class="mdescLeft">&#160;</td><td class="mdescRight">parameters supposed to be set once  <a href="#adb95e5cb1cbb2c43a1889131cb017c2b">More...</a><br /></td></tr>
<tr class="separator:adb95e5cb1cbb2c43a1889131cb017c2b"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Function Documentation</h2>
<a id="abca3b5ef9a0cde5e39c94287e35b31da"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abca3b5ef9a0cde5e39c94287e35b31da">&#9670;&nbsp;</a></span>annotate_df()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.annotate_df </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>row</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>df</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>min_dist</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>ax</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>create annotation label for a point on axis if it's far enough from other points used internally by plotting functions </p>

</div>
</div>
<a id="a8f980392d707316c84323ff4da652e62"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8f980392d707316c84323ff4da652e62">&#9670;&nbsp;</a></span>find_pseudotime()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.find_pseudotime </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pca</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>function </p>
<ul>
<li>finds pair of 2 PCs that are most correlated with time labels (as defined by "day" column in annotation table) using spearman correlation</li>
<li>finds rotation of this PCs so X axis has best correlation with time</li>
</ul>
<p>returns: pseudotime for each cell, defined as linear combination of PCs, having best time correlation</p>
<p>arguments are:</p><ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>annotation pd.DataFrame</li>
<li>pca sklearn.decomposition object</li>
<li>settings object </li>
</ul>

</div>
</div>
<a id="aa11b06d3030a58a0287811a14e00de0c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#aa11b06d3030a58a0287811a14e00de0c">&#9670;&nbsp;</a></span>get_isoforms_correlated_with_pc()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.get_isoforms_correlated_with_pc </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>expression_table</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pc</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>n</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>save genes correlated with a PC to file </p>

</div>
</div>
<a id="a2dd8534ee3983dc13dac25891db005fe"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a2dd8534ee3983dc13dac25891db005fe">&#9670;&nbsp;</a></span>main()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.main </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>main function when run separately, program expects following arguments: </p>
<ul>
<li>argv[1] = comma separated file with expression values</li>
<li>argv[2] = file with cell sets (see <a class="el" href="classsc-analyses_1_1settings.html#a1339c4a1025a673ff4a4538734676895" title="read sets of cells later used to refer to them (to remove/color/superimpose etc...) ">settings.read_cell_sets()</a>) </li>
</ul>

</div>
</div>
<a id="af8827644de4784f9b91bd82a316ba4c7"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af8827644de4784f9b91bd82a316ba4c7">&#9670;&nbsp;</a></span>plot_2d_pca_multiplot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.plot_2d_pca_multiplot </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pca</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>create plot of 6 PC combinations PC1 vs PC2, PC3 vs PC4 etc. </p>
<p>arguments are:</p><ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>annotation pd.DataFrame</li>
<li>pca sklearn.decomposition object</li>
<li>settings object </li>
</ul>

</div>
</div>
<a id="abbc100d6dc158e3b5a1fbba583da3e10"></a>
<h2 class="memtitle"><span class="permalink"><a href="#abbc100d6dc158e3b5a1fbba583da3e10">&#9670;&nbsp;</a></span>plot_2d_pca_single_plot()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.plot_2d_pca_single_plot </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pca</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot cells of defined pair of PCs arguments are: </p>
<ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>annotation pd.DataFrame</li>
<li>pca sklearn.decomposition object</li>
<li>settings object </li>
</ul>

</div>
</div>
<a id="a7ce99634f6c5778fe1ffb3a6333840f8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a7ce99634f6c5778fe1ffb3a6333840f8">&#9670;&nbsp;</a></span>plot_3d_pca()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.plot_3d_pca </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>height</em> = <code>1080</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>width</em> = <code>1600</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>create 3d PCA plot using plotly library arguments are: </p>
<ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>annotation pd.DataFrame</li>
<li>settings object </li>
</ul>

</div>
</div>
<a id="a8f5f78fd44190040bfde554a412bceaa"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8f5f78fd44190040bfde554a412bceaa">&#9670;&nbsp;</a></span>plot_gene_with_pseudotime()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.plot_gene_with_pseudotime </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>exp</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>pseudotime</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transcript_id</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plots gene expression over pseudotime arguments are: </p>
<ul>
<li>pd.DataFrame with gene expression</li>
<li>pd.Series with pseudotime coordinates for each cell</li>
<li>Ensamble transcript ID </li>
</ul>

</div>
</div>
<a id="ae3c71f2d759657520fbffb935d9f727c"></a>
<h2 class="memtitle"><span class="permalink"><a href="#ae3c71f2d759657520fbffb935d9f727c">&#9670;&nbsp;</a></span>plot_hierarchycal_clusterings()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.plot_hierarchycal_clusterings </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>plot hierarchycal clustering arguments are: </p>
<ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>annotation pd.DataFrame</li>
<li>settings object</li>
<li>filename for output picture </li>
</ul>

</div>
</div>
<a id="a96e3d0a6c4cbd66f450cd13d0597d693"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a96e3d0a6c4cbd66f450cd13d0597d693">&#9670;&nbsp;</a></span>read_expression()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.read_expression </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>expression_file</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>settings</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>min_expression</em> = <code>0.1</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>min_cells</em> = <code>10</code>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>log_transform</em> = <code>True</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>function takes expression file and settings object and returns: </p>
<ul>
<li>pd.DataFrame with [log transformed] expression values [genes expressed over min_expression in at least min_cells]</li>
<li>pd.DataFrame with annotations for each cell. Expression table and annotation table have the same rows </li>
</ul>

</div>
</div>
<a id="a48155c82bbcbbab0bf4a6b4ae399252d"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a48155c82bbcbbab0bf4a6b4ae399252d">&#9670;&nbsp;</a></span>rotate_expression()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.rotate_expression </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>transformed_expression</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>x</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>y</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>angle</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>rotate transformed expression matrix by defined angle used internally in order to define pseudotime arguments are: </p>
<ul>
<li>pd.DataFrame with PCA transformed gene expression</li>
<li>x,y = PCs to rotate</li>
<li>angle in degrees returns: pdDataFrame with values in columns x,y rotated by angle </li>
</ul>

</div>
</div>
<a id="a8c5c34b3ded1400d6294ac2263f02c52"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a8c5c34b3ded1400d6294ac2263f02c52">&#9670;&nbsp;</a></span>run_PCA()</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">def sc-analyses.run_PCA </td>
          <td>(</td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>expression_table</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>annotation</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">&#160;</td>
          <td class="paramname"><em>n_components</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>runs PCA and returns: </p>
<ul>
<li>PCA transformed coordinates</li>
<li>sklearn.decomposition.pca object </li>
</ul>

</div>
</div>
<h2 class="groupheader">Variable Documentation</h2>
<a id="adb95e5cb1cbb2c43a1889131cb017c2b"></a>
<h2 class="memtitle"><span class="permalink"><a href="#adb95e5cb1cbb2c43a1889131cb017c2b">&#9670;&nbsp;</a></span>accepted_parameters</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list sc-analyses.accepted_parameters = [&quot;number_of_genes&quot;]</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>parameters supposed to be set once </p>

</div>
</div>
<a id="a43a788e866dfeecd5963f4ee948d4008"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a43a788e866dfeecd5963f4ee948d4008">&#9670;&nbsp;</a></span>accepted_sets_with_parameter</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list sc-analyses.accepted_sets_with_parameter = [&quot;color&quot;, &quot;outline-color&quot;, &quot;size&quot;, &quot;name&quot;, &quot;shape&quot;, &quot;superimpose&quot;, &quot;superimpose-for-spearman&quot;]</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sets with parameter look like: operation set_name parameter for ex. </p>
<p>: color day_4 clue </p>

</div>
</div>
<a id="af635bfeee6b34ad984fe2512c83d6be3"></a>
<h2 class="memtitle"><span class="permalink"><a href="#af635bfeee6b34ad984fe2512c83d6be3">&#9670;&nbsp;</a></span>accepted_sets_without_parameter</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list sc-analyses.accepted_sets_without_parameter = [&quot;remove&quot;]</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sets without parameter look like: operation set_name for ex. </p>
<p>: remove low_read_count_cells </p>

</div>
</div>
<a id="a5c4820a68409c2ca9d5965d48c1643c8"></a>
<h2 class="memtitle"><span class="permalink"><a href="#a5c4820a68409c2ca9d5965d48c1643c8">&#9670;&nbsp;</a></span>run_modes</h2>

<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">list sc-analyses.run_modes = [&quot;find_day_correlated_pcs&quot;, &quot;2d-pca-multiplot&quot;, &quot;2d-pca-single&quot;, &quot;3d-pca&quot;, &quot;hierarchy&quot;, &quot;pseudotime&quot;]</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>what modes can be script run in </p>

</div>
</div>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
