#!/usr/bin Rscript

args <- (commandArgs(trailingOnly = TRUE))
for (i in seq_len(length(args))) {
  eval(parse(text = args[[i]]))
}

# Rprof(rprof_out)

print(paste0("num_clones: ", init_k))

library(numbat)
library(Seurat)
library(readr)
library(magrittr)
conflicted::conflict_prefer("rowSums", "Matrix")

print(matrix_file)

matrix_dir = fs::path_dir(matrix_file)

count_mat <- Seurat::Read10X(matrix_dir)

# subset count matrix by cell to ease tree construction ------------------------------
cell_ceiling <- as.numeric(cell_ceiling)
if(cell_ceiling > 0){

	cell_ceiling = ifelse(ncol(count_mat) > cell_ceiling, cell_ceiling, ncol(count_mat))

	count_mat <- count_mat[,sample(ncol(count_mat),size=cell_ceiling,replace=FALSE)]
}

# test dropping read numbers ------------------------------
if(!as.numeric(read_prop) == 1){
	count_mat <- scuttle::downsampleMatrix(count_mat, as.numeric(read_prop))
}

allele_df = data.table::fread(allele_df) %>%
    tidyr::drop_na(CHROM) %>% # drop X chromosome values
    # dplyr::mutate(CHROM = as.factor(CHROM)) %>%
  identity()

# assemble a normal reference from non-RB cell types ------------------------------
bad_cell_types = c("RPCs", "Late RPCs", c("Red Blood Cells", "Microglia", "Muller Glia", "RPE", "Horizontal Cells", "Rod Bipolar Cells", "Pericytes", "Bipolar Cells", "Astrocytes", "Endothelial", "Schwann", "Fibroblasts"))

# myseu <- readRDS(seu_path)
#
# num_out_cells <- sum(myseu$type %in% bad_cell_types)

# if(num_out_cells > 100){
# 	normal_seu = myseu[,myseu$type %in% bad_cell_types]
#
# 	normal_reference_mat <-
# 		normal_seu %>%
# 		GetAssayData(slot = "counts")
#
# 	normal_cell_annot <-
# 		normal_seu@meta.data[c("type")] %>%
# 		tibble::rownames_to_column("cell") %>%
# 		dplyr::rename(group = type) %>%
# 		identity()
#
# 	ref_internal = numbat::aggregate_counts(normal_reference_mat, normal_cell_annot)
# } else {
# 	ref_internal <- readRDS(ref_path)
# }

print(ref_path)
ref_internal <- readRDS(ref_path)

bulk = numbat:::get_bulk(
	count_mat = count_mat,
	lambdas_ref = ref_internal,
	df_allele = allele_df,
	gtf = gtf_hg38
)

segs_loh = bulk %>% numbat:::detect_clonal_loh(t = as.numeric(t))

# segs_loh = NULL

# run
out = numbat::run_numbat(
	count_mat, # gene x cell integer UMI count matrix done
	lambdas_ref = ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix done
	allele_df, # allele dataframe generated by pileup_and_phase script done
	genome = "hg38",
	min_cells = 10,
	t = as.numeric(t),
	gamma = as.numeric(gamma),
	max_iter = as.integer(max_iter),
	max_entropy = max_entropy,
	segs_loh = segs_loh,
	min_LLR = min_LLR,
	init_k = init_k,
	ncores = as.integer(ncores),
	plot = TRUE,
	multi_allelic = FALSE,
	common_diploid = TRUE,
	out_dir = out_dir,
	tau = as.numeric(tau),
	skip_nj = TRUE,
	expression_only = FALSE)

done_file <- fs::path(out_dir, "done.txt")

write(ref_path, file = done_file)
