#!/usr/bin Rscript

args <- (commandArgs(trailingOnly = TRUE))
for (i in seq_len(length(args))) {
  eval(parse(text = args[[i]]))
}

# Rprof(rprof_out)

library(numbat)
library(Seurat)
library(readr)
library(magrittr)
conflicted::conflict_prefer("rowSums", "Matrix")
library(glue)

cell_ceiling = 1e4
read_prop = 1
t = 1e-5
max_iter = 2
max_entropy = 0.5
min_LLR = 5
ncores = 4
tau = 0.3
init_k = 3
  
study = "wu"
sample_id = "SRR13884244"
out_dir = glue("output/numbat/{study}_et_al/{sample_id}")

matrix_dir = glue("~/single_cell_projects/resources/{study}_et_al_proj/output/cellranger/{sample_id}/outs/filtered_feature_bc_matrix/")

allele_df = glue("~/single_cell_projects/resources/{study}_et_al_proj/output/numbat/{sample_id}/{sample_id}_allele_counts.tsv.gz")

count_mat <- Seurat::Read10X(matrix_dir)

# subset count matrix by cell to ease tree construction ------------------------------
cell_ceiling <- as.numeric(cell_ceiling)
if(cell_ceiling > 0){
	
	cell_ceiling = ifelse(ncol(count_mat) > cell_ceiling, cell_ceiling, ncol(count_mat))
	
	count_mat <- count_mat[,sample(ncol(count_mat),size=cell_ceiling,replace=FALSE)]
}

# test dropping read numbers ------------------------------
if(!as.numeric(read_prop) == 1){
	count_mat <- scuttle::downsampleMatrix(count_mat, as.numeric(read_prop))
}

allele_df = data.table::fread(allele_df) %>%
  tidyr::drop_na(CHROM) %>% # drop X chromosome values
  dplyr::mutate(CHROM = as.factor(CHROM)) %>%
  identity()

ref_path <- "~/Homo_sapiens/numbat/collin_ref.rds"

ref_internal <- readRDS(ref_path)

bulk = numbat:::get_bulk(
	count_mat = count_mat,
	lambdas_ref = ref_internal,
	df_allele = allele_df,
	gtf = gtf_hg38
)

segs_loh = bulk %>% numbat:::detect_clonal_loh(t = as.numeric(t))

# segs_loh = NULL

# run
# debug(numbat::run_numbat)

# step into iteration 2; get_exp_post is most important function! ------------------------------
# numbat::get_exp_post
# numbat:::get_exp_likelihoods()

# iteration 0
hc <- readRDS(glue("{out_dir}/hc.rds"))
subtrees = numbat:::get_nodes_celltree(hc, cutree(hc, k = init_k))

# iteration 1
subtrees <- readRDS(glue("{out_dir}/subtrees_1.rds"))
clones <- readRDS(glue("{out_dir}/clones_1.rds"))

gtf = gtf_hg38

rstudioapi::filesPaneNavigate(out_dir)

debug(run_numbat)

out = run_numbat(
	count_mat, # gene x cell integer UMI count matrix done
	lambdas_ref = ref_internal, # reference expression profile, a gene x cell type normalized expression level matrix done
	allele_df, # allele dataframe generated by pileup_and_phase script done
	genome = "hg38",
	min_cells = 10,
	t = as.numeric(t),
	max_iter = as.integer(max_iter),
	max_entropy = max_entropy,
	segs_loh = segs_loh,
	min_LLR = min_LLR,
	init_k = 3,
	ncores = as.integer(ncores),
	plot = TRUE,
	multi_allelic = FALSE,
	common_diploid = TRUE,
	out_dir = out_dir,
	tau = as.numeric(tau),
	skip_nj = FALSE)

nb = Numbat$new(out_dir = out_dir)

nb_path = fs::path(out_dir, paste0(fs::path_file(out_dir), "_numbat.rds"))

saveRDS(nb, nb_path)

